src/lib.rs

```diff
@@ -59,4 +63,57 @@ impl Calc {
             .collect::<Vec<String>>();
         return msgs.join("\n");
     }
+
+    pub fn eval(opcode: AstNode) -> Result<u64, String> {
+        let bytecode = &mut vec![];
+        Self::to_bytecode(opcode, bytecode);
+        return Self::eval_bytecode(bytecode);
+    }
+
+    pub fn to_bytecode(ast_node: AstNode, result: &mut Vec<Instruction>) -> &Vec<Instruction> {
+        match ast_node {
+            AstNode::Add { lhs, rhs } => {
+                Self::to_bytecode(*lhs, result);
+                Self::to_bytecode(*rhs, result);
+                result.push(Instruction::Add {})
+            }
+            AstNode::Mul { lhs, rhs } => {
+                Self::to_bytecode(*lhs, result);
+                Self::to_bytecode(*rhs, result);
+                result.push(Instruction::Mul {})
+            }
+            AstNode::Number { value } => result.push(Instruction::Push { value: value }),
+        }
+        return result;
+    }
+
+    pub fn eval_bytecode(instructions: &Vec<Instruction>) -> Result<u64, String> {
+        let mut stack: Stack<u64> = Stack::new();
+        for a in instructions {
+            match a {
+                Instruction::Push { value } => stack.push(*value),
+                Instruction::Mul {} => {
+                    let result = stack
```

Comment:
you should be able to get these two lines more compact with something like:

```
stack.push(stack.pop().checked_mul(...).ok_or(...)?;
```

Reply:
I'm not sure how the error from the `pop` function will be returned with this example. Can you explain?

Reply:
Your `pop()` wrapper doesn't return any error, so you have nothing to handle.

Do you mean the error from `checked_mul()`? The`?` operator means "unwrap or return error".

Reply:
oops, it's definitely not `pop`... 
changed :point_right:  https://github.com/softdevteam/pavel.yaiwr/pull/10/commits/8e960eada019da4a18def6435a73c1fa200a2f0e

Reply:
You can even compress the two lines into one :)

Reply:
when I try as one line:
```rust
                    stack.push(stack
                        .pop()
                        .checked_mul(stack.pop())
                        .ok_or("overflowed".to_string())?)
```

gives me this error: 
```
error[E0499]: cannot borrow `stack` as mutable more than once at a time
  --> src/lib.rs:91:38
   |
89 |                       stack.push(stack
   |                       -     ---- first borrow later used by call
   |  _____________________|
   | |
90 | |                         .pop()
91 | |                         .checked_mul(stack.pop())
   | |                                      ^^^^^^^^^^^ second mutable borrow occurs here
92 | |                         .ok_or("overflowed".to_string())?)
   | |__________________________________________________________- first mutable borrow occurs her
```

Reply:
Ah, the compiler requires it as two lines.

What's funny is that they should be semantically equivalent.

Oh well. Leave it as two lines.